### PR TITLE
[EMCS-272] add message validation framework

### PR DESCRIPTION
**Pull request by mpbostock**
To be picked up by someone else, while I'm away

### CODE REVIEW

File: app/uk/gov/hmrc/excisemovementcontrolsystemapi/controllers/DraftExciseMovementController.scala [hunk 1]
```diff
@@ -45,6 +45,8 @@ class DraftExciseMovementController @Inject()(
   extends BackendController(cc) {
 
   def submit: Action[NodeSeq] =
+
+  //TODO EMCS-400 - remove the validateErnInMessageAction and instead use the MessageValidation to find the correct Ern
```
Comment (mauro-fanton) [inline app/uk/gov/hmrc/excisemovementcontrolsystemapi/controllers/DraftExciseMovementController.scala:49]: can we do this now,  so we can rid of of this todo?

File: app/uk/gov/hmrc/excisemovementcontrolsystemapi/controllers/DraftExciseMovementController.scala [hunk 2]
```diff
@@ -59,7 +61,11 @@ class DraftExciseMovementController @Inject()(
     }
 
   private def submitMessage(boxId: String)(implicit request: ValidatedXmlRequest[NodeSeq]): Future[Result] = {
-    submissionMessageService.submit(request).flatMap {
+    // TODO when using MessageValidation will have one ern returned after validation
+    // TODO For now, we know the consignor is fine because it passed the validate action above.
+    val submittingErn = request.message.asInstanceOf[IE815Message].consignorId
```
Comment (mauro-fanton) [inline app/uk/gov/hmrc/excisemovementcontrolsystemapi/controllers/DraftExciseMovementController.scala:66]: can we use MessageValidation now as part of this work? At least for the draft Movement, so we can rid off of this cast.

File: app/uk/gov/hmrc/excisemovementcontrolsystemapi/controllers/SubmitMessageController.scala [hunk 3]
```diff
@@ -30,32 +39,80 @@ import scala.xml.NodeSeq
...
+        } yield {
+          Accepted
         }
 
+        result.merge
+    }
+
+  }
+
+  private def validateMovementId(movementId: String): EitherT[Future, Result, Movement] = {
```
Comment (mauro-fanton) [inline app/uk/gov/hmrc/excisemovementcontrolsystemapi/controllers/SubmitMessageController.scala:69]: Can we re-use this for getMessageController too? We could have a MovementIdValidator or similar to be reused in both places

File: app/uk/gov/hmrc/excisemovementcontrolsystemapi/models/messages/IEMessage.scala [hunk 4]
```diff
@@ -35,4 +37,13 @@ trait IEMessage {
   def toXml: NodeSeq
 
   def lrnEquals(lrn: String): Boolean
+}
+
+object IEMessage {
```
Comment (mauro-fanton) [inline app/uk/gov/hmrc/excisemovementcontrolsystemapi/models/messages/IEMessage.scala:42]: I would have this as a trait and make the IE837Message and the IE871Message derived from this. Putting this as an object ,as companion of the IEMessage suggest all the messages should have this, while with the trait we make a distinct separation:

`
trait SubmitterTypeConverter {
  def convert(submitterType: SubmitterType): ExciseTraderType = {
    submitterType match {
      case Number1Value31 => Consignor
      case Number2Value30 => Consignee
    }
  }
}


case class IE837Message
(
  private val obj: IE837Type,
  private val key: Option[String],
  private val namespace: Option[String]
) **extends IEMessage with SubmitterTypeConverter** {

  def submitter: ExciseTraderType = **convert**(obj.Body.ExplanationOnDelayForDelivery.AttributesValue.SubmitterType)
  def consignorId: Option[String] = submitter match {
    case Consignor => Some(obj.Body.ExplanationOnDelayForDelivery.AttributesValue.SubmitterIdentification)
    case _ => None
  }
.....
}

`

File: app/uk/gov/hmrc/excisemovementcontrolsystemapi/services/MovementService.scala [hunk 5]
```diff
@@ -63,6 +65,19 @@ class MovementService @Inject()(
     movementRepository.getMovementById(id)
   }
 
+  def validateMovementId(id: String): Future[Either[MovementValidationError, Movement]] = {
```
Comment (mauro-fanton) [inline app/uk/gov/hmrc/excisemovementcontrolsystemapi/services/MovementService.scala:68]: I would rename this to getMovementId and call the movementRepository.getMovementById(id) here and delete the current getMovementId. I Know things function does two things (violate the single responsibility principle) but we will always validate the Id before getting the movementId from the repository so I think this is ok. There are few places where we validate the Id and then get the movementId from db.

File: app/uk/gov/hmrc/excisemovementcontrolsystemapi/controllers/DraftExciseMovementController.scala (hunk 1 above)
Comment (ChrisCookOC) [reply app/uk/gov/hmrc/excisemovementcontrolsystemapi/controllers/DraftExciseMovementController.scala]: done

File: app/uk/gov/hmrc/excisemovementcontrolsystemapi/controllers/DraftExciseMovementController.scala (hunk 2 above)
Comment (ChrisCookOC) [reply app/uk/gov/hmrc/excisemovementcontrolsystemapi/controllers/DraftExciseMovementController.scala]: done

File: app/uk/gov/hmrc/excisemovementcontrolsystemapi/controllers/SubmitMessageController.scala (hunk 3 above)
Comment (ChrisCookOC) [reply app/uk/gov/hmrc/excisemovementcontrolsystemapi/controllers/SubmitMessageController.scala:67]: done

File: app/uk/gov/hmrc/excisemovementcontrolsystemapi/models/messages/IEMessage.scala (hunk 4 above)
Comment (ChrisCookOC) [reply app/uk/gov/hmrc/excisemovementcontrolsystemapi/models/messages/IEMessage.scala]: done

File: app/uk/gov/hmrc/excisemovementcontrolsystemapi/services/MovementService.scala (hunk 5 above)
Comment (ChrisCookOC) [reply app/uk/gov/hmrc/excisemovementcontrolsystemapi/services/MovementService.scala]: I've moved this validateMovementId to its own class

File: app/uk/gov/hmrc/excisemovementcontrolsystemapi/models/validation/MovementIdValidation.scala [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+import java.time.Instant
+import java.util.UUID
+import javax.inject.Inject
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.Try
+
+case class MovementIdValidation @Inject()(movementService: MovementService)(implicit ec: ExecutionContext) {
+
+  def validateMovementId(id: String): Future[Either[MovementValidationError, Movement]] = {
```
Comment (mauro-fanton) [inline app/uk/gov/hmrc/excisemovementcontrolsystemapi/models/validation/MovementIdValidation.scala:34]: I would not call the movementService.getMovementId from here as this function does not just do the validation, but it retrieve the Id too. I would move the validation inside movementService.getMovementId or call the validator from movementService.getMovementId.

**Review by mauro-fanton — APPROVED** [review]

**Pull request merged**